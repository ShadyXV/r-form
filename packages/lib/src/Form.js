// @flow
import React, { Component } from 'react';

export const { Provider, Consumer } = React.createContext();

type Props = {
  onSubmit: () => boolean,
  onChange: (string) => any,
};

class Form extends Component <Props> {
  state = this.props.value || {};

  nodes = [];

  register(name) {
    return (node) => {
      if (node === null) {
        this.nodes = this.nodes.filter(n => n.name === name);
      } else {
        this.nodes = this.nodes.concat({ name, node });
      }
    };
  }

  get(name) {
    return () => this.state[name];
  }

  update(name) {
    return (text) => {
      this.setState((prevState) => {
        const { onChange } = this.props;
        const v = onChange && onChange(name, text, prevState, this);
        if (typeof v === 'object') {
          return {
            [name]: text,
            ...v,
          };
        }
        return { [name]: text };
      });
      this.nodes.filter(n => n.name === name).forEach(n => n.node.validate(text));
    };
  }

  next(name) {
    return () => {
      const idx = this.nodes.findIndex(n => n.name === name) + 1;
      if (idx < this.nodes.length) {
        this.nodes[idx].node.focus();
      }
    };
  }

  submit() {
    const { state, props } = this;
    this.nodes.forEach((iNode) => {
      const { node } = iNode;
      node.validate(node.props.value);
    });
    // props.onSubmit(state);
  }

  render() {
    const {
      onSubmit, onChange, value, ...other
    } = this.props;
    this.register = this.register.bind(this);
    this.get = this.get.bind(this);
    this.next = this.next.bind(this);
    this.update = this.update.bind(this);
    this.submit = this.submit.bind(this);
    return (
      <Provider value={{ form: this, state: this.state }} {...other} />
    );
  }
}

export default Form;
