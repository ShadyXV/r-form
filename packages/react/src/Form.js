import React, { Component } from 'react';
import { TextInput } from 'react-native';

const { Provider, Consumer } = React.createContext();

class Form extends Component {
  state = {}
  nodes = [];

  register(name) {
    return (node) => {
      if (node === null) {
        this.nodes = this.nodes.filter(n => n.name === name);
      } else {
        this.nodes = this.nodes.concat({ name, node });
      }
    }
  }

  get(name) {
    return this.state[name];
  }

  update(name) {
    return (text) => {
      this.setState((prevState) => {
        const { onChange } = this.props;
        const v = onChange && onChange(name, text, prevState, this);
        if (typeof v === 'object') {
          return {
            [name]: text,
            ...v,
          };
        }

        return { [name]: text };
      });
    };
  }

  next(name) {
    return () => {
      console.log('Next called', name);
      const idx = this.nodes.findIndex(n => n.name === name) + 1;
      if (idx < this.nodes.length) {
        this.nodes[idx].node.focus();
      }
    }
  }

  submit() {
    this.props.onSubmit(this.state);
  }

  render() {
    const { onSubmit, onChange, ...other } = this.props;
    return (
      <Provider value={{ form: this, state: this.state}} {...other} />
    );
  }
}

class Input extends Component {
  shouldComponentUpdate(nextProps) {
    const check = nextProps.value !== this.props.value;
    return check;
  }

  render() {
    const { children, ...form } = this.props;
    return children(form);
  }
}

Form.Input = ({ name, ...other }) => (
  <Consumer>
    {({ form, state }) => (
      <Input
        register={form.register(name)}
        value={form.get(name)}
        update={form.update(name)}
        next={form.next(name)}
        {...other}
      />
    )}
  </Consumer>
);

Form.Consumer = Consumer;

export default Form;



