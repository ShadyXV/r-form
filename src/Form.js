// @flow
import React, { Component, Fragment } from 'react';
import Input from './Input';
import FormArray from './FormArray';

export const { Provider, Consumer } = React.createContext({});

class Form extends Component {
  state = this.props.value || {};

  nodes = [];

  register(name) {
    return (node) => {
      if (node === null) {
        this.nodes = this.nodes.filter(n => n.name === name);
      } else {
        this.nodes = this.nodes.concat({ name, node });
      }
    };
  }

  get(name) {
    return () => this.state[name];
  }

  update(name) {
    return (text) => {
      this.setState((prevState) => {
        const { onChange } = this.props;
        const v = onChange && onChange(name, text, prevState, this);
        if (typeof v === 'object') {
          return {
            [name]: text,
            ...v,
          };
        }
        return { [name]: text };
      });
      this.nodes.filter(n => n.name === name).forEach(n => n.node.validate(text));
    };
  }

  next(name) {
    return () => {
      const idx = this.nodes.findIndex(n => n.name === name) + 1;
      if (idx < this.nodes.length) {
        this.nodes[idx].node.focus();
      }
    };
  }

  submit() {
    const { state, props } = this;
    this.nodes.forEach((iNode) => {
      const { node } = iNode;
      node.validate(node.props.value);
    });
    props.onSubmit(state);
  }

  render() {
    const {
      onSubmit, onChange, value, ...other
    } = this.props;
    return (
      <Provider value={{ form: this, state: this.state }} {...other} />
    );
  }
}

Form.Input = ({
  name, validator, visibility, ...other
}) => (
  <Consumer>
    {({ form, state }) => (
      <Input
        validator={validator}
        state={state}
        ref={form.register(name)}
        get={form.get(name)}
        update={form.update(name)}
        next={form.next(name)}
        visibility={visibility}
        {...other}
      />
    )}
  </Consumer>
);

Form.Array = FormArray;

const groupOnChange = (name, value, prevState) => ({
  ...prevState,
  [name]: value,
});

Form.Group = ({ name, ...other }) => (
  <Form.Input name={name} {...other}>
    {form => (
      <Fragment>
        <Form
          value={form.value}
          onChange={(n, value, prevState) => form.update(groupOnChange(n, value, prevState))}
          onSubmit={() => form.submit()}
          {...other}
        />
      </Fragment>
    )}
  </Form.Input>
);

Form.Consumer = Consumer;
export default Form;
