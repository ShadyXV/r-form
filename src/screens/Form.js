import React, { Component, Fragment } from 'react';
import Input from './Input';

const { Provider, Consumer } = React.createContext({});

class Form extends Component {
  state = this.props.value || {};

  nodes = [];

  register(name) {
    return (node) => {
      if (node === null) {
        this.nodes = this.nodes.filter(n => n.name === name);
      } else {
        this.nodes = this.nodes.concat({ name, node });
      }
    };
  }

  get(name) {
    return this.state[name];
  }

  update(name) {
    return (text) => {
      this.setState((prevState) => {
        const { onChange } = this.props;
        const v = onChange && onChange(name, text, prevState, this);
        if (typeof v === 'object') {
          return {
            [name]: text,
            ...v,
          };
        }
        return { [name]: text };
      });
      this.nodes.filter(n => n.name === name).forEach(n => n.node.validate(text));
    };
  }

  next(name) {
    return () => {
      console.log('Next called', name);
      const idx = this.nodes.findIndex(n => n.name === name) + 1;
      if (idx < this.nodes.length) {
        this.nodes[idx].node.focus();
      }
    };
  }

  submit() {
    const { state, props } = this;
    this.nodes.forEach((iNode) => {
      const { node } = iNode;
      node.validate(node.props.value);
    });
    props.onSubmit(state);
  }

  render() {
    const {
      onSubmit, onChange, value, ...other
    } = this.props;
    return (
      <Provider value={{ form: this, state: this.state }} {...other} />
    );
  }
}

Form.Input = ({
  name, validator, visibility, ...other
}) => (
  <Consumer>
    {({ form, state }) => (
      <Input
        validator={validator}
        state={state}
        ref={form.register(name)}
        value={form.get(name)}
        update={form.update(name)}
        next={form.next(name)}
        visibility={visibility}
        {...other}
      />
    )}
  </Consumer>
);

const arrayOnChange = (name, value, prevState, idx) => {
  console.log('change');
};

Form.Array = ({ name, children, onChange, auto, ...other }) => (
  <Form.Input name={name} {...other}>
    {form => (
      <Fragment>
        {form.value.map((v, idx) => children(name))}
      </Fragment>
    )}
  </Form.Input>
);


const groupOnChange = (name, value, prevState) => ({
  ...prevState,
  [name]: value,
});

Form.Group = ({ name, ...other }) => (
  <Form.Input name={name} {...other}>
    {form => (
      <div>
        <Form
          value={form.value}
          onChange={(n, value, prevState) => form.update(groupOnChange(n, value, prevState))}
          onSubmit={() => form.submit()}
          {...other}
        />
      </div>
    )}
  </Form.Input>
);

Form.Consumer = Consumer;
export default Form;
