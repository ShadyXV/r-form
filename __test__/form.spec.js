import React, { Fragment } from 'react';
import { mount } from 'enzyme';
import Form from '../src';
import './setupTests';

const value = {
  name: 'Blah',
  spouse: { name: 'spousename', age: 11 },
  spouses: [{ name: 'A', age: 20 }, { name: 'B', age: 21 }, { name: 'C', age: 22 }],
  hobby: ['Drawing'],
  nameList: ['Nun', 'Another', 'One'],
  something: [['1', '2']],
};

const hobbies = ['Drawing', 'Sports', 'Art'];

const DomForm = ({ onSubmit, onChange, ...props }) => (
  <form onSubmit={() => onSubmit()} onChange={() => onChange()}>
    <Form {...props} />
  </form>
);


const InputBox = ({ name, value, validator}) => (
  <Form.Input name={name}>
    {(form) => {
      return (
        <input
          name={name}
          value={form.get()}
          onChange={(e) => {
            e.preventDefault();
            form.update(e.target.value);
          }}
        />
      );
    }}
  </Form.Input>
);

const FormArray = ({ name, InputType }) => (
  <Form.Array name={name} InputType={InputType} />
);

const FormGroup = ({ name }) => (
  <Form.Group name={name}>
    <InputBox name="name" />
    <InputBox name="age" />
  </Form.Group>
);


const CheckBox = ({ name, domain, ...other }) => (
  <Form.Input name={name} {...other}>
    {form => (domain.map(d => (
      <Fragment key={d}>
        <input
          type="checkbox"
          value={d}
          checked={form.get().includes(d)}
          onChange={e => (
            e.target.checked ? form.update(form.value.concat(e.target.value)) : form.update(form.value.filter(v => v !== e.target.value))
          )}
        />
      </Fragment>
    ))

    )}
  </Form.Input>
);

const App = ({ onSubmit, onChange }) => (
  <DomForm onSubmit={() => onSubmit()} onChange={() => onChange()} value={value}>
    {/* <InputBox name="name" /> <br /> */}
    <FormArray name="nameList" InputType={InputBox} />
    {/* <FormGroup name="spouse" /> */}
    {/* <CheckBox name="hobby" domain={hobbies} /> */}
    {/* <FormArray name="spouses" InputType={FormGroup} /> */}

    {/* array within array */}
    <FormArray name="something" InputType={() => <FormArray name="name" InputType={InputBox} />} />
  </DomForm>
);

const changeFn = jest.fn();
const submitFn = jest.fn();
const component = mount(<App onSubmit={() => submitFn()} onChange={() => changeFn()} />);

describe('Form component Test', () => {
  test('Form Input on Change', () => {
    expect(component).toMatchSnapshot();

    // const input = component.find('input').first();
    // input.simulate('change', { target: { value: 'new Value' } });

    // const inp = component.find('input').first();

    // expect(inp.props().value).toBe('new Value');

    // expect(changeFn).toHaveBeenCalled();
  });

  // test('Form Group on Change', () => {
  //   const formGroup = component.find(FormGroup).first();
  //   const firstInputBox = formGroup.find('input').first();

  //   firstInputBox.simulate('change', { target: { value: 'new Value' } });

  //   const changeformGroup = component.find(Form).first();
  //   const changeInputBox = changeformGroup.find('input').first();
  //   expect(changeInputBox.props().value).toBe('new Value');
  // });

  // test('Form Array on Change', () => {
  //   const formArray = component.find(FormArray).first();
  //   const firstInputBox = formArray.find('input').first();

  //   firstInputBox.simulate('change', { target: { value: 'change nun' } });
  //   const changedFormArray = component.find(FormArray).first();
  //   const changedInputBox = changedFormArray.find('input').first();

  //   expect(changedInputBox.props().value).toBe('change nun');
  // });

  // test('Form Array and Form Group', () => {
  //   const formArray = component.find(FormArray).last();
  //   const formGroup = formArray.find(FormGroup).first();
  //   const inputBox = formGroup.find('input').first();

  //   inputBox.simulate('change', { target: { value: 'change name' } });

  //   const changedFormArray = component.find(FormArray).last();
  //   const changedFormGroup = changedFormArray.find(FormGroup).first();
  //   const changedInputBox = changedFormGroup.find('input').first();
  //   console.log(changedFormArray.props().value);
  //   expect(changedInputBox.props().value).toBe('change name');
  // });
});
