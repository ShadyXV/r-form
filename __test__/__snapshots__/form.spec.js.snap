// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component Test Form Input on Change 1`] = `
<App
  onChange={[Function]}
  onSubmit={[Function]}
>
  <DomForm
    onChange={[Function]}
    onSubmit={[Function]}
    value={
      Object {
        "hobby": Array [
          "Drawing",
        ],
        "name": "Blah",
        "nameList": Array [
          "Nun",
          "Another",
          "One",
        ],
        "something": Array [
          Array [
            "1",
            "2",
          ],
        ],
        "spouse": Object {
          "age": 11,
          "name": "spousename",
        },
        "spouses": Array [
          Object {
            "age": 20,
            "name": "A",
          },
          Object {
            "age": 21,
            "name": "B",
          },
          Object {
            "age": 22,
            "name": "C",
          },
        ],
      }
    }
  >
    <form
      onChange={[Function]}
      onSubmit={[Function]}
    >
      <Form
        value={
          Object {
            "hobby": Array [
              "Drawing",
            ],
            "name": "Blah",
            "nameList": Array [
              "Nun",
              "Another",
              "One",
            ],
            "something": Array [
              Array [
                "1",
                "2",
              ],
            ],
            "spouse": Object {
              "age": 11,
              "name": "spousename",
            },
            "spouses": Array [
              Object {
                "age": 20,
                "name": "A",
              },
              Object {
                "age": 21,
                "name": "B",
              },
              Object {
                "age": 22,
                "name": "C",
              },
            ],
          }
        }
      >
        <FormArray
          InputType={[Function]}
          name="nameList"
        >
          <Array
            InputType={[Function]}
            name="nameList"
          >
            <Input
              name="nameList"
            >
              <InputHelper
                get={[Function]}
                next={[Function]}
                state={
                  Object {
                    "hobby": Array [
                      "Drawing",
                    ],
                    "name": "Blah",
                    "nameList": Array [
                      "Nun",
                      "Another",
                      "One",
                    ],
                    "something": Array [
                      Array [
                        "1",
                        "2",
                      ],
                    ],
                    "spouse": Object {
                      "age": 11,
                      "name": "spousename",
                    },
                    "spouses": Array [
                      Object {
                        "age": 20,
                        "name": "A",
                      },
                      Object {
                        "age": 21,
                        "name": "B",
                      },
                      Object {
                        "age": 22,
                        "name": "C",
                      },
                    ],
                  }
                }
                update={[Function]}
              >
                <InputBox
                  name="name"
                >
                  <Input
                    name="name"
                  >
                    <InputHelper
                      get={[Function]}
                      next={[Function]}
                      state={
                        Array [
                          "Nun",
                          "Another",
                          "One",
                        ]
                      }
                      update={[Function]}
                    >
                      <input
                        name="name"
                        onChange={[Function]}
                      />
                    </InputHelper>
                  </Input>
                </InputBox>
              </InputHelper>
            </Input>
          </Array>
        </FormArray>
      </Form>
    </form>
  </DomForm>
</App>
`;
